src/plugin/systems/mod.rs

```diff
@@ -3,65 +3,74 @@
 mod character_controller;
 mod collider;
 mod joint;
+mod multiple_rapier_contexts;
 mod remove;
 mod rigid_body;
 mod writeback;
 
 pub use character_controller::*;
 pub use collider::*;
 pub use joint::*;
+pub use multiple_rapier_contexts::*;
 pub use remove::*;
 pub use rigid_body::*;
 pub use writeback::*;
 
 use crate::dynamics::{RapierRigidBodyHandle, TransformInterpolation};
 use crate::pipeline::{CollisionEvent, ContactForceEvent};
 use crate::plugin::configuration::SimulationToRenderTime;
-use crate::plugin::{RapierConfiguration, RapierContext};
+use crate::plugin::{RapierConfiguration, RapierContext, TimestepMode};
 use crate::prelude::{BevyPhysicsHooks, BevyPhysicsHooksAdapter};
 use bevy::ecs::system::{StaticSystemParam, SystemParamItem};
 use bevy::prelude::*;
 
 /// System responsible for advancing the physics simulation, and updating the internal state
 /// for scene queries.
 pub fn step_simulation<Hooks>(
-    mut context: ResMut<RapierContext>,
-    config: Res<RapierConfiguration>,
+    mut context: Query<(
+        &mut RapierContext,
+        &RapierConfiguration,
+        &mut SimulationToRenderTime,
+    )>,
+    timestep_mode: Res<TimestepMode>,
     hooks: StaticSystemParam<Hooks>,
     time: Res<Time>,
-    mut sim_to_render_time: ResMut<SimulationToRenderTime>,
-    collision_events: EventWriter<CollisionEvent>,
-    contact_force_events: EventWriter<ContactForceEvent>,
-    interpolation_query: Query<(&RapierRigidBodyHandle, &mut TransformInterpolation)>,
+    mut collision_events: EventWriter<CollisionEvent>,
+    mut contact_force_events: EventWriter<ContactForceEvent>,
+    mut interpolation_query: Query<(&RapierRigidBodyHandle, &mut TransformInterpolation)>,
 ) where
     Hooks: 'static + BevyPhysicsHooks,
     for<'w, 's> SystemParamItem<'w, 's, Hooks>: BevyPhysicsHooks,
 {
-    let context = &mut *context;
     let hooks_adapter = BevyPhysicsHooksAdapter::new(hooks.into_inner());
 
-    if config.physics_pipeline_active {
-        context.step_simulation(
-            config.gravity,
-            config.timestep_mode,
-            Some((collision_events, contact_force_events)),
-            &hooks_adapter,
-            &time,
-            &mut sim_to_render_time,
-            Some(interpolation_query),
-        );
-        context.deleted_colliders.clear();
-    } else {
-        context.propagate_modified_body_positions_to_colliders();
-    }
+    for (mut context, config, mut sim_to_render_time) in context.iter_mut() {
+        let context = &mut *context;
+
+        if config.physics_pipeline_active {
+            context.step_simulation(
```

Comment:
Perhaps in future we could look to par_itering this loop, since this is the most expensive part of the physics engine and could now be easily split across multiple worlds.

Reply:
probably ! the difficult part is reconciliating the events which is the same eventWriter for every contexts.

Again I'd like Entity events :/ 

In the meantime it might be possible to get `collision_events_to_send` and `contact_force_events_to_send` out of RapierContext, out of this loop, so we're populating directly the `rwlock<vec<events>>` ; and read it only once all the worlds have been simulated.

I'll keep that as a follow up if needed. (Please remind me to or create an issue if that's something highly desirable :) )